crates/ty_python_semantic/src/types/class.rs

```diff
@@ -883,27 +884,122 @@ impl<'db> ClassLiteral<'db> {
             )
             .symbol;
 
-        if let Symbol::Type(Type::BoundMethod(metaclass_call_function), _) =
-            metaclass_call_function_symbol
+        if let Symbol::Type(Type::BoundMethod(metaclass_dunder_call_function), _) =
+            metaclass_dunder_call_function_symbol
         {
             // TODO: this intentionally diverges from step 1 in
             // https://typing.python.org/en/latest/spec/constructors.html#converting-a-constructor-to-callable
             // by always respecting the signature of the metaclass `__call__`, rather than
             // using a heuristic which makes unwarranted assumptions to sometimes ignore it.
-            return Some(metaclass_call_function.into_callable_type(db));
+            return metaclass_dunder_call_function.into_callable_type(db);
         }
 
-        let dunder_new_method = self_ty
-            .find_name_in_mro(db, "__new__")
-            .expect("find_name_in_mro always succeeds for class literals")
-            .symbol
-            .try_call_dunder_get(db, self_ty);
+        let dunder_new_function_symbol = self_ty
+            .member_lookup_with_policy(
+                db,
+                "__new__".into(),
+                MemberLookupPolicy::MRO_NO_OBJECT_FALLBACK
+                    | MemberLookupPolicy::META_CLASS_NO_TYPE_FALLBACK,
+            )
+            .symbol;
+
+        let dunder_new_function =
+            if let Symbol::Type(Type::FunctionLiteral(dunder_new_function), _) =
+                dunder_new_function_symbol
+            {
+                // Step 3: If the return type of the `__new__` evaluates to a type that is not a subclass of this class,
+                // then we should ignore the `__init__` and just return the `__new__` method.
+                let new_return_ty = match dunder_new_function.signature(db) {
+                    FunctionSignature::Single(signature)
+                    | FunctionSignature::Overloaded(_, Some(signature)) => signature.return_ty,
+                    FunctionSignature::Overloaded(..) => None,
+                };
+
+                let dunder_new_bound_method =
+                    dunder_new_function.into_bound_method_type(db, self_ty);
+
+                if let Some(new_return_ty) = new_return_ty {
+                    if !new_return_ty.to_meta_type(db).is_subtype_of(db, self_ty) {
+                        return dunder_new_bound_method;
+                    }
+                }
+                Some(dunder_new_bound_method)
+            } else {
+                None
+            };
+
+        let dunder_init_function_symbol = self_ty
+            .member_lookup_with_policy(
+                db,
+                "__init__".into(),
+                MemberLookupPolicy::MRO_NO_OBJECT_FALLBACK
+                    | MemberLookupPolicy::META_CLASS_NO_TYPE_FALLBACK,
+            )
+            .symbol;
+
+        // TODO: should be the concrete value of `Self`
```

Comment:
I'm not actually sure what the spec means by "the concrete value of `Self`" here; I don't know what else that could be, other than what you have here (for non-generic classes, anyway).